### PR TITLE
docs: translate API Reference / `react` / Hooks

### LINKED ISSUE #397: react: Hooks
Page: https://id.react.dev/reference/react

Assignee: @KevinPalowa

### PR DESCRIPTION
**Pull request by KevinPalowa**
translation for react: hooks page
issue #330 
Closes #397

### CODE REVIEW

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:31]: Ada yang kurang:

```suggestion
*Context* memungkinkan sebuah komponen [menerima informasi dari induk yang jauh tanpa mengopernya sebagai props.](/learn/passing-props-to-a-component) Sebagai contoh, komponen tingkat atas aplikasi Anda dapat meneruskan tema antarmuka pengguna (UI) saat ini ke semua komponen di bawah, tidak peduli seberapa dalam. 
```

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:43]: Timeout di sini refer ke setTimeout API dari JavaScript, sebaiknya gak usah di terjemahkan.

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:46]: Expose = ekspos
```suggestion
* [`useImperativeHandle`](/reference/react/useImperativeHandle) memungkinkan Anda menyesuaikan *ref* yang terekspos oleh komponen Anda. Ini jarang sekali digunakan.
```

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:56]: Effect di sini refer ke API nya React.
Gak usah diterjemahkan apa ya? 
CC: @mhaidarhanif @resir014 @r17x @zainfathoni

Comment (r17x) [reply src/content/reference/react/index.md:56]: "Effect" dalam tulisan ini konteksnya grouping dan juga istilah. 

Menurutku masih okay untuk ditranslate kecuali API Efek seperti `useEffect`

Comment (KevinPalowa) [reply src/content/reference/react/index.md:56]: Need to confirm which one should i use

Comment (mazipan) [reply src/content/reference/react/index.md:56]: Then it's okay to translate

Comment (KevinPalowa) [reply src/content/reference/react/index.md:56]: thx for the confirmation, i'll make this conversation resolve then

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:101]: Typo 

```suggestion
- [`useTransition`](/reference/react/useTransition) memungkinkan Anda menandai transisi state sebagai non-blocking dan mengizinkan pembaruan lain untuk menghentikannya.
```

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:102]:
```suggestion
- [`useDeferredValue`](/reference/react/useDeferredValue) memungkinkan Anda menangguhkan pembaruan bagian antarmuka pengguna (UI) yang tidak penting dan membiarkan bagian lain memperbarui terlebih dahulu.
```

File: src/content/reference/react/index.md
Comment (mazipan) [inline src/content/reference/react/index.md:97]:
```suggestion
Terkadang, Anda tidak dapat melewatkan pe-*render*an ulang karena layar sebenarnya perlu pembaruan. Di dalam kasus itu, Anda dapat meningkatkan performa dengan memisahkan memblokir pembaruan yang harus sinkron (seperti mengetik masukan) dari pembaruan non-pemblokiran yang tidak perlu memblokir antarmuka pengguna (seperti memperbarui bagan).
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:83]:
```suggestion
Cara yang umum untuk mengoptimalkan kinerja *rendering* ulang adalah melewatkan pekerjaan yang tidak perlu. Sebagai contoh, Anda dapat memberi tahu React untuk menggunakan kembali perhitungan yang sudah ada di dalam *cache* atau untuk melewatkan *render* ulang jika data tersebut tidak berubah sejak render sebelumnya.
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:85]:
```suggestion
Untuk melewatkan perhitungan dan *render* ulang yang tidak perlu, gunakan salah satu dari Hook ini:
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:87]:
```suggestion
- [`useMemo`](/reference/react/useMemo) memungkinkan Anda menyimpan hasil perhitungan yang berat secara *cache*.
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:88]:
```suggestion
- [`useCallback`](/reference/react/useCallback) memungkinkan Anda menyimpan definisi fungsi di *cache* sebelum melewatkannya ke komponen yang dioptimalkan.
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:99]:
```suggestion
Untuk memprioritaskan *rendering*, gunakan salah satu Hook ini:
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:101]:
```suggestion
- [`useTransition`](/reference/react/useTransition) memungkinkan Anda menandai transisi *state* sebagai non-blocking dan mengizinkan pembaruan lain untuk menghentikannya.
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:16]:
```suggestion
Untuk menambahkan *state* ke sebuah komponen, gunakan salah satu Hooks dibawah ini:
```

File: src/content/reference/react/index.md
Comment (r17x) [inline src/content/reference/react/index.md:97]:
```suggestion
Terkadang, Anda tidak dapat melewatkan *rendering* ulang karena layar sebenarnya perlu pembaruan. Di dalam kasus itu, Anda dapat meningkatkan performa dengan memisahkan memblokir pembaruan yang harus sinkron (seperti mengetik masukan) dari pembaruan non-pemblokiran yang tidak perlu memblokir antarmuka pengguna (seperti memperbarui bagan).
```

**Review by r17x — CHANGES_REQUESTED** [review]

**Review by r17x — CHANGES_REQUESTED** [review]
Kata seperti "browser" dan mungkin beberapa kata lain belum ditranslate. 

* Browser -> peramban
* browser DOM -> DOM Peramban

**Comment by KevinPalowa** [comment]
> Kata seperti "browser" dan mungkin beberapa kata lain belum ditranslate.
> 
> * Browser -> peramban
> * browser DOM -> DOM Peramban

noted,

Butuh masukan untuk kata `repaint` sebaiknya menggunakan kata apa ya?

**Comment by mhaidarhanif** [comment]
@KevinPalowa `repaint` can be written with just `*repaint*` or like ` mengecat ulang (*repaint*)`

File: src/content/reference/react/index.md
Comment (mhaidarhanif) [inline src/content/reference/react/index.md:76]: @KevinPalowa `repaint` can be written with just `*repaint*` (italicized) or like ` mengecat ulang (*repaint*)`

Need feedback @resir014 @r17x @mazipan @zainfathoni

**Review by mhaidarhanif — CHANGES_REQUESTED** [review]
Suggestion.

File: src/content/reference/react/index.md
Comment (mhaidarhanif) [inline src/content/reference/react/index.md:76]: @KevinPalowa `repaint` can be written with just `*repaint*` (italicized) or like `melukis ulang (*repaint*)`

Need feedback @resir014 @r17x @mazipan @zainfathoni 

```suggestion
* [`useLayoutEffect`](/reference/react/useLayoutEffect) dipanggil sebelum browser melukis ulang (*repaint*) layar. Anda dapat menggunakannya untuk mengukur layout pada tahap ini.
```

**Review by mhaidarhanif — CHANGES_REQUESTED** [review]
Improvement, inspired by @zainfathoni suggestion: https://github.com/reactjs/id.react.dev/discussions/337#discussioncomment-5768813

File: src/content/reference/react/index.md
Comment (KevinPalowa) [reply src/content/reference/react/index.md:76]: need confirmation is it ok to use `melukis ulang`

Comment (zainfathoni) [reply src/content/reference/react/index.md:76]: Yes, it is okay. And I think we should also either translate `browser` to `peramban` or at least make it italic *browser*.

**Review by r17x — APPROVED** [review]

File: src/content/reference/react/index.md
Comment (zainfathoni) [inline src/content/reference/react/index.md:97]:
```suggestion
Terkadang, Anda tidak dapat melewatkan pe-*render*-an ulang karena layar sebenarnya perlu pembaruan. Di dalam kasus itu, Anda dapat meningkatkan kinerja dengan memisahkan pembaruan penghalang yang harus sinkron (seperti mengetik masukan) dari pembaruan yang bukan penghalang yang tidak perlu memblokir antarmuka pengguna (seperti memperbarui bagan).
```

Comment (zainfathoni) [reply src/content/reference/react/index.md:97]: Shall we add "blocking" as "*penghalang*" into the glossary, @resir014 @r17x @mhaidarhanif @mazipan?

File: src/content/reference/react/index.md
Comment (zainfathoni) [inline src/content/reference/react/index.md:99]:
```suggestion
Untuk memprioritaskan pe-*render*-an, gunakan salah satu dari *Hook-hook* ini:
```

File: src/content/reference/react/index.md
Comment (zainfathoni) [inline src/content/reference/react/index.md:101]:
```suggestion
- [`useTransition`](/reference/react/useTransition) memungkinkan Anda menandai transisi *state* sebagai bukan penghalang dan mengizinkan pembaruan lain untuk menghentikannya.
```

**Review by zainfathoni — CHANGES_REQUESTED** [review]
I have more suggestions, mostly related to translating "*rendering*" into "pe-*render*-an". 😁 

I also have suggestions to introduce new terms into the glossary, @resir014 @r17x @mhaidarhanif @mazipan 

| English        | Indonesian       |
| -------------- | ---------------- |
| blocking (verb)          | menghalangi          | ❓ |
| blocking (noun)          | penghalang          | ❓ |
| blocking update          | pembaruan penghalang          | ❓ |
| non-blocking update          | pembaruan yang bukan penghalang          | ❓ |

Let's discuss them here: https://github.com/reactjs/id.react.dev/discussions/337#discussioncomment-5772140

**Review by mazipan — APPROVED** [review]

File: src/content/reference/react/index.md
Comment (zainfathoni) [inline src/content/reference/react/index.md:78]:
```suggestion
* [`useInsertionEffect`](/reference/react/useInsertionEffect) dipanggil sebelum React melakukan perubahan ke DOM. Pustaka-pustaka (*libraries*) lainnya dapat memasukkan CSS dinamik di sini.
```

Missing the closing asterisk `*`.

image

File: src/content/reference/react/index.md
Comment (zainfathoni) [inline src/content/reference/react/index.md:98]:
```suggestion
Terkadang, Anda tidak dapat melewatkan pe-*render*-an ulang karena layar memang benar-benar memerlukan pembaruan. Di dalam kasus itu, Anda dapat meningkatkan kinerja dengan memisahkan pembaruan penghalang (*blocking updates*) yang harus sinkron (seperti mengetik masukan) dari pembaruan yang bukan penghalang (*non-blocking updates*) yang tidak perlu memblokir antarmuka pengguna (seperti memperbarui bagan).
```

1. Translate "actually needs" to "memang benar-benar memerlukan" to make it more explicit.
2. After re-reading it in bahasa Indonesia, I think we need to mention the original "*blocking updates*" and "*non-blocking updates*" terms to introduce the translated words.

**Review by zainfathoni — CHANGES_REQUESTED** [review]
I have more minor suggestions for my second review. Please address them. Thanks!

**Comment by r17x** [comment]
Need re-review @zainfathoni

File: src/content/reference/react/index.md
Comment (resir014) [inline src/content/reference/react/index.md:73]:
```suggestion
Efek adalah "jalan keluar" dari paradigma React. Jangan gunakan Efek untuk mengatur aliran data aplikasi Anda. Jika Anda tidak berinteraksi dengan sistem eksternal, [kemungkinan besar Anda tidak memerlukan Efek.](/learn/you-might-not-need-an-effect)
```

- We can put "jalan keluar" inside quotemarks just like the original text.
- Missing period.

**Review by resir014 — CHANGES_REQUESTED** [review]
One minor change. @KevinPalowa

**Comment by mhaidarhanif** [comment]
@resir014 and @zainfathoni can check and approve if okay.

**Review by resir014 — APPROVED** [review]
All good. Thanks!

**Pull request merged**
